graphql_compiler/compiler/compiler_frontend.py

```diff
@@ -457,14 +476,20 @@ def _compile_vertex_ast(schema, current_schema_type, ast,
                 basic_blocks.append(blocks.Backtrack(location, optional=True))
 
                 # Exiting optional block!
-                # Add a tag right after the optional, to ensure future Backtrack blocks
+                # Add a MarkLocation right after the optional, to ensure future Backtrack blocks
                 # return to a position after the optional set of blocks.
                 location = location.revisit()
                 context['location_types'][location] = strip_non_null_from_type(current_schema_type)
                 basic_blocks.append(_mark_location(location))
+                marked_location_stack.pop()
+                marked_location_stack.append({'location': location, 'num_traverses': 0})
             else:
                 basic_blocks.append(blocks.Backtrack(location))
 
+    # Pop off the initial MarkLocation for the current vertex.
+    if not is_in_fold_scope(context):
+        marked_location_stack.pop()
```

Comment:
For bonus points, here's a sanity check that I think would be nice to have:
- Record the length of the stack upon entering this function.
- Then, right after this `.pop()`, check that the current stack length is equal to the initial stack length at the start of the function. Since the comment here says we're popping off the initial location we pushed earlier, the lengths must always match.